index.d.ts

```diff
@@ -1,10 +1,44 @@
-import algoliasearch from 'algoliasearch';
+// Note: Bellow, we will be importing both algoliasearch
+// `v3` and algoliasearch `v4` types. The goal is being
+// able to export the algoliasearch-helper types using
+// the developer installed version of the client.
+
+import algoliasearch, {
+  // @ts-ignore
+  SearchClient as SearchClientV4,
+  // @ts-ignore
+  Client as SearchClientV3,
+  // @ts-ignore
+  QueryParameters as SearchOptionsV3,
+  // @ts-ignore
+  Response as SearchResponseV3,
+} from 'algoliasearch';
+import {
+  SearchOptions as SearchOptionsV4,
+  SearchResponse as SearchResponseV4
+// @ts-ignore
+} from '@algolia/client-search';
 import { EventEmitter } from 'events';
 
-type SearchClient = Pick<
-  algoliasearch.Client,
-  'search' | 'searchForFacetValues'
->;
+type DummySearchClientV4 = {
+  readonly addAlgoliaAgent: (segment: string, version?: string) => void;
+};
+
+type Client = ReturnType<typeof algoliasearch> extends DummySearchClientV4
+  ? SearchClientV4
+  : SearchClientV3;
+type SearchOptions = ReturnType<
```

Comment:
There's no way to reuse the `Client` type?

Something like:

```ts
type SearchOptions = Client extends SearchClientV4
  ? SearchOptionsV4
  : SearchOptionsV3;
```

Reply:
`SearchClientV4` may be any when is not installed. Everything extends any. So in your example `SearchOptions` will be any too.

Reply:
I think @francoischalifour meant: `type SearchOptions = Client extends DummySearchClientV4`